Mathlib/RingTheory/Localization/FractionRingAlgebra.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang, Yakov Pechersky
+-/
+import Mathlib.RingTheory.Localization.Integral
+import Mathlib.RingTheory.RingHom.Finite
+
+/-! # Instances about extensions of fraction rings -/
+
+variable {R₀ R S : Type*} [CommRing R] [CommRing S] [Algebra R S] [FaithfulSMul R S]
+  [IsDomain S] [CommRing R₀] [Algebra R₀ R] [Algebra R₀ S] [IsScalarTower R₀ R S]
+
+open scoped nonZeroDivisors
+
+namespace FractionRing
+
+noncomputable section
+
+/-- This causes a diamond for `Algebra (FractionRing R) (FractionRing (FractionRing R))`
+but we will hardly ever see `FractionRing (FractionRing R)` in mathlib. -/
+instance : Algebra (FractionRing R) (FractionRing S) :=
+  -- this cannot be an instance because it wouldn't be able to synthesize S
+  haveI : IsDomain R := (FaithfulSMul.algebraMap_injective R S).isDomain
+  FractionRing.liftAlgebra _ _
+
+lemma algebraMap_fractionRing_eq_map :
+    algebraMap (FractionRing R) (FractionRing S) =
+      IsFractionRing.map (FaithfulSMul.algebraMap_injective R S) :=
+  rfl
+
+instance : FaithfulSMul (FractionRing R) (FractionRing S) := by
```

Comment:
Maybe it's possible to use [IsLocalization.lift_injective_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Localization/Defs.html#IsLocalization.lift_injective_iff)
This could be an instance about `liftAlgebra`.

Reply:
factored out and upstreamed